### PR TITLE
List.append() returns index of inserted element

### LINKED ISSUE #17466: Should/could `List.append()` return the index of the element that was inserted?
### Summary of Problem

As a Chapel programmer, it seems that it would be convenient for `List.append()` to return the list index where the element was inserted, particularly in the case of multiple tasks inserting items into the list simultaneously (such that querying the list's size before or after insertion wouldn't be guaranteed to be atomic w.r.t. the appended element).

### Steps to Reproduce

**Source Code:**

```chapel
use List;
var l: list(int);
var i = l.append(42);
writeln(l[i]);
```

**Compile command:**
`chpl testit.chpl`

### Configuration Information

- Output of `chpl --version`: `1.24.0`

### PR DESCRIPTION
**Pull request by strikeraryu**
I have added the functionality to return the index of the inserted element Introduced in Issue #17466. Currently, I have made changes in List.chpl.
Resolves #17466

Further changes can be made regarding documentation and testing.

### CODE REVIEW

**Comment by bradcray** [comment]
@strikeraryu : You should also add a new test that locks in this change to make sure it doesn't break in the future.  This can be done in `test/library/standard/List/`.  See https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/TestSystem.rst for more information on the Chapel test system if you haven't used it before.

**Comment by strikeraryu** [comment]
Should I make the test for both serial and parallel cases?

**Comment by bradcray** [comment]
> Should I make the test for both serial and parallel cases?

Good question.  I think that would be ideal, though writing deterministic parallel tests can be challenging.  Though perhaps an approach would just be to have a number of tasks call insert() with a unique integer simultaneously, store the index they got back into a normal array, and then have a serial task after the parallel section run through that array verifying that the expected integer is at the expected list index?

**Comment by strikeraryu** [comment]
I wrote a test to check manually for parallel before (mentioned in issue #17466) 
``` chapel
coforall i in 0..10 with (ref l){
    var ind = l.append(i);
    writeln(ind, " ", i);
}
writeln(l);
```
can I update this to make it automated?

**Comment by bradcray** [comment]
Ideally, you wouldn't want to print within the parallel loop, or print out the list elements since the output order will be nondeterministic across runs, so would be hard to embed in a .good file within the test system (without post-processing).  This is why I was suggesting storing the `ind` values into an array indexed by `i` where the array is declared outside the loop. 
 Then you can check after the loop that `l[arr[i]] == i;` for each i in 0..10, printing success or failures/assertion errors and having the .good file just contain the success output (since it should always be true).

**Comment by strikeraryu** [comment]
Yeah, that was for just checking manually. I will automate this by the ways you mentioned, Thanks.

**Comment by strikeraryu** [comment]
I have added the test for both serial and parallel, Should I update the documentation with the changes done in ```list.append()```.

**Comment by dlongnecke-cray** [comment]
Updating the documentation would be great, thanks!

**Comment by strikeraryu** [comment]
What are the main files concerning the documentation?

**Comment by dlongnecke-cray** [comment]
You'll want to update the docstring on `list.append()` to mention that it returns the indice of the appended element. You can check your changes to the docs with something like:

```
cd $CHPL_HOME
make docs
cd doc/html
python3 -m http.server
```

Then in your browser head to `localhost:8000`, navigate to `Standard Modules` -> `List`, and check out how your changes look in the rendered documentation for `list.append()`.

**Comment by bradcray** [comment]
> You'll want to update the docstring on list.append()

Just in case this isn't crystal clear to @strikeraryu (it wouldn't have been to me a few years ago), the documentation is generated from the comments that precede the routine in the source code.  The `make docs` step that David outlines above will generate the docs from the source.  I don't personally tend to spin up a server to view them, but just point a browser at `$CHPL_HOME/docs/html/index.html` and go from there (as simple as `open docs/html/index.html` on a Mac).

**Comment by strikeraryu** [comment]
I am getting error while ```make docs```, is there a recent change related to this, as my local branch is behind.

**Comment by bradcray** [comment]
> I am getting error while make docs, is there a recent change related to this, as my local branch is behind.

There shouldn't be... what kind of error?

**Comment by strikeraryu** [comment]
I dont remember the error exactly but it was somethin related to make error in man and then there was no html file in doc folder. I doing some changes, if i get same error again I send the exact error.

**Comment by bradcray** [comment]
Hmm, that one is not at all familiar to me, and I'm not much of a Python guru, which is where the problem seems to be.  This might be worth opening a new issue for, or asking about on Discourse or Gitter, as I believe it's independent of your effort here.  My first reaction is "Could something be wrong with your Python installation?  Could re-installing Python fix it?" but that's a shot in the dark.

Since it looks to be specific to making the man pages, and you're not working there at the moment, it might be worth editing your copy of `Makefile.devel` to remove the `man-*` dependencies in the `docs:` rule and just focus on the chpldoc docs.  (Of course, if something's wrong with your Python setup, you may hit different issues, but that could be interesting as well).

**Comment by strikeraryu** [comment]
> it might be worth editing your copy of Makefile.devel to remove the man-* dependencies in the docs: rule and just focus on the chpldoc docs.

while trying out this I got this error
image

**Comment by strikeraryu** [comment]
It finally worked for me, and I think we are good to go. I edited the docstring for ```List.append()``` when is first edited the function but now I have tested and it is giving the right details.
image

**Comment by bradcray** [comment]
@dlongnecke-cray : Do you have time to review this PR due to your familiarity with the List module?

**Comment by dlongnecke-cray** [comment]
Yes, I will review it.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [inline modules/standard/List.chpl:709]: Let's go ahead and rename `insertedElementIndex` to `result` to stay consistent with the rest of the module.

**Review by dlongnecke-cray — CHANGES_REQUESTED** [review]
Go ahead and check the updated docs locally to make sure they look OK.

In the OP, mention:

- What testing you've done (e.g., you checked the docs, maybe ran the tests in `test/library/standard/List`
- Add a "Resolves #17466" so that the issue will auto close after this pull is merged

I'll check out the branch and start running tests. Thanks!

**Comment by bradcray** [comment]
> Go ahead and check the updated docs locally to make sure they look OK.

David: Doesn't the screen shot above already demonstrate that they do?

**Comment by dlongnecke-cray** [comment]
Haha that's what I get for only looking on the `Files` tab!

**Comment by dlongnecke-cray** [comment]
Hi @strikeraryu, there were some test failures. Mind taking a look?

```
[Test Summary - 210407.223805]
[Error matching compiler output for classes/delete-free/lifetimes/bad-append-borrow]
[Error matching compiler output for library/standard/List/classLists/listAppendBorrow]
[Error matching compiler output for library/standard/List/classLists/listAppendLifetime]
[Error matching program output for library/standard/List/listAppendLifetime]
[Summary: #Successes = 12599 | #Failures = 5 | #Futures = 0]
[END]
```

**Comment by bradcray** [comment]
@dlongnecke-cray / @strikeraryu : I was surprised that anything would fail under this PR, so checked the branch out to reproduce and am seeing that we're not getting the same lifetime errors from `append()` as we did prior to this change.  I'm trying to guess whether that implies there's a pre-existing bug here, or whether something about making the method return changes the semantics or abilities of the lifetime checker.  Tagging @mppf in case it's more obvious to him that to us.

(sample diffs for anyone not wanting to bother building the branch:  The lifetime errors embedded in the `.good` files due to the append() calls are not present in the output now, e.g.:

```diff
[Executing diff listAppendBorrow.good listAppendBorrow.comp.out.tmp]
1,5d0
< listAppendBorrow.chpl:16: In function 'error7':
< listAppendBorrow.chpl:17: error: Actual argument does not meet called function lifetime constraint
< listAppendBorrow.chpl:16: note: consider scope of arg
< $CHPL_HOME/modules/standard/List.chpl:nnnn: note: called function list(borrowed foo?,false).append(x: borrowed foo?)
< $CHPL_HOME/modules/standard/List.chpl:nnnn: note: includes lifetime constraint this < x
```

**Comment by mppf** [comment]
> I'm trying to guess whether that implies there's a pre-existing bug here, or whether something about making the method return changes the semantics or abilities of the lifetime checker. Tagging @mppf in case it's more obvious to him that to us.

Unfortunately this sounds like a bug in the lifetime checker. The `lifetime this < x` constraint should cause the error, no matter the return type.

**Comment by strikeraryu** [comment]
I was going through this and I was getting the same error, so started looking about lifeline in Chapel, I was not able to understand the problem because I am neither returning a reference nor storing a variable with local scope into a variable with a larger scope.

**Comment by dlongnecke-cray** [comment]
I will see if I can find a reproducer to fix this bug/look for a quick fix.

**Comment by dlongnecke-cray** [comment]
I haven't quite found a fix yet, so my plan is to go ahead and merge your work @strikeraryu and then deal with the fallout later by adding a future or two. My rationale is that that the test failures arguably uncovered a bug that we wouldn't have found otherwise (so extra props in that regard)!

So sometime tomorrow I'll be hitting the merge button 👍.

**Comment by strikeraryu** [comment]
Great, you can tell me if you want me for anything, Thanks.

**Comment by dlongnecke-cray** [comment]
Thanks @strikeraryu!

**Comment by bradcray** [comment]
@dlongnecke-cray : Weird... when I `git log` this PR, I get the log message:

```
Merge pull request #17510 from strikeraryu/master

Added a const ref overload to the methods first() and last() using
helpers. Previously the methods did not accept the usage of const
ref intent types.

Resolves #17259

-

Contributed by @strikeraryu, reviewed by @dlongnecke-cray.

[Reviewer comments follow...]

Tested with COMM=none and COMM=gasnet runs. Checked docs locally.

Several lifetime checker tests related to `list` will fail as a
consequence of merging this PR. They are not a result of this effort
but rather a bug that changing the return type of `list.append()`
from `void` to `int` has uncovered.

I will open a PR addressing test failures and adding futures as a
separate effort.

Thanks @strikeraryu!
```

cut-and-paste error?

**Comment by dlongnecke-cray** [comment]
No, that is the merge commit exactly as I formatted it. The lines above the `-` are intended to be the user's merge commit exactly as they wrote it. The lines below the `-` are my comments.

**Comment by bradcray** [comment]
Isn't this the PR that made list.append() return the index of the new element?  Looking at the file diffs, first/last seem untouched, and the OP says:

```
List.append() returns index of inserted element

I have added the functionality to return the index of the inserted element Introduced in Issue #17466. Currently, I have made changes in List.chpl.
Resolves #17466

Further changes can be made regarding documentation and testing.
```

**Comment by dlongnecke-cray** [comment]
I'm timing out from thinking about closures today. You're correct that the message is a typo. I must have copied the wrong OP without checking it twice.

**Comment by bradcray** [comment]
No problem (and I don't think it's worth fixing), I just noticed it while surfing through git logs and writing up status reports.

**Comment by strikeraryu** [comment]
Should I mention this in OP so if someone comes looking in the future he realizes quickly that it was just a typo, or is it ok as it is?

**Comment by bradcray** [comment]
Thanks for asking, @strikeraryu.  I think we're good with this comment trail.

**Pull request merged**
